### PR TITLE
webgpu: Make fromPixels not so special

### LINKED ISSUE #5536: webgpu: fromPixels for HTMLVideoElement related cases fail to execute 'importExternalTexture'
Run `yarn test` under tfjs-backend-webgpu.
Expect all cases pass. However, below cases fail:
```
FAILED fromPixels test-webgpu {"WEBGPU_CPU_FORWARD":false} fromPixels for HTMLVideoElement
debug.js:21 Error: Failed to execute 'importExternalTexture' on 'GPUDevice': Failed to import texture from video
error properties: Object({ code: 0, INDEX_SIZE_ERR: 1, DOMSTRING_SIZE_ERR: 2, HIERARCHY_REQUEST_ERR: 3, WRONG_DOCUMENT_ERR: 4, INVALID_CHARACTER_ERR: 5, NO_DATA_ALLOWED_ERR: 6, NO_MODIFICATION_ALLOWED_ERR: 7, NOT_FOUND_ERR: 8, NOT_SUPPORTED_ERR: 9, INUSE_ATTRIBUTE_ERR: 10, INVALID_STATE_ERR: 11, SYNTAX_ERR: 12, INVALID_MODIFICATION_ERR: 13, NAMESPACE_ERR: 14, INVALID_ACCESS_ERR: 15, VALIDATION_ERR: 16, TYPE_MISMATCH_ERR: 17, SECURITY_ERR: 18, NETWORK_ERR: 19, ABORT_ERR: 20, URL_MISMATCH_ERR: 21, QUOTA_EXCEEDED_ERR: 22, TIMEOUT_ERR: 23, INVALID_NODE_TYPE_ERR: 24, DATA_CLONE_ERR: 25 })
    at Object.fromPixelsExternalImage (FromPixelsExternalImage.ts:84)
    at Object.fromPixels [as kernelFunc] (FromPixels.ts:65)
```
However, if I change flag `WEBGPU_USE_IMPORT` to false, above case can pass. It seems that the import path has   problem.
cc @shaoboyan

### PR DESCRIPTION
**Pull request by gyagp**
This is to unify the implementation of fromPixels with other ops for better maintenance. 

Close #5536

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6549)

### CODE REVIEW

**Comment by gyagp** [comment]
@qjia7 @axinging @xhcao PTAL!

**Review by qjia7 — APPROVED** [review]
LGTM, thanks!

**Comment by gyagp** [comment]
@qjia7 @axinging @xhcao I did further refactoring to unify the texture and buffer. Please review this again.

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 1]
```diff
@@ -729,94 +720,47 @@ export class WebGPUBackend extends KernelBackend {
...
-            util.getTypedArrayFromDType(output.dtype as 'float32', 0);
-        return output;
-      }
-      this.uploadToGPU(output.dataId);
     }
+    if (util.sizeFromShape(output.shape) === 0) {
+      // Short-circuit the computation since the result is empty (has 0 in its
+      // shape).
+      const outData = this.tensorMap.get(output.dataId);
+      outData.values =
```
[the remark this answers is not in the document]
Comment (gyagp) [reply tfjs-backend-webgpu/src/backend_webgpu.ts:726]: All the tensor should be recorded in tensorMap. If output is not null, then it's already recorded.

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 2]
```diff
@@ -604,10 +586,12 @@ export class WebGPUBackend extends KernelBackend {
     }
 
     const tensorData = this.tensorMap.get(tensor.dataId);
-
+    if (tensorData.textureInfo) {
+      return (tensorData.textureInfo.texture as GPUTexture).createView();
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:590]: Forget the situation that `tensorData.textureInfo.texture` may be a `GPUExternalTexture`?

File: tfjs-backend-webgpu/src/kernels/FromPixels.ts [hunk 3]
```diff
@@ -60,14 +60,23 @@ export function fromPixels(args: {
...
   if (isImageBitmap || isCanvas || isVideoOrImage) {
-    let externalImage;
+    let texture: GPUTexture|GPUExternalTexture;
+    let textureInfo: TextureInfo;
     if (useImport) {
-      externalImage = pixels as HTMLVideoElement;
+      const externalTextureDescriptor = {source: pixels as HTMLVideoElement};
+      texture = backend.device.importExternalTexture(externalTextureDescriptor);
+      textureInfo = {
+        width: pixels.width,
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/FromPixels.ts:74]: Can you directly use the `width` `height` variables? `pixels.width` and `pixels.height` may not equal to `width` `height` in outputShape (see L57).

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 4]
```diff
@@ -900,172 +833,24 @@ export class WebGPUBackend extends KernelBackend {
...
     this.queue.copyExternalImageToTexture(
         {source: externalImage}, {texture}, [outShape[1], outShape[0]]);
 
-    const textureInfo = {
-      width: outShape[1],
-      height: outShape[0],
-      format: textureFormat,
-      usage: textureUsage,
+    return {
+      width: externalImage.width,
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:848]: Be careful that `externalImage.width` may be not equal to `outShape[1]`

**Review by qjia7 — COMMENTED** [review]
Please also check whether all fromPixels cases pass if set `WEBGPU_USE_IMPORT` to true.

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 5]
```diff
@@ -195,23 +190,17 @@ export class WebGPUBackend extends KernelBackend {
 
   flushDisposalQueue() {
     this.tensorDisposalQueue.forEach(d => {
-      this.maybeReleaseBuffer(d);
+      this.releaseResource(d);
       this.tensorMap.delete(d);
     });
     this.uniformDisposalQueue.forEach(
-        d => this.bufferManager.releaseBuffer(d.buffer, d.byteSize, d.usage));
+        d => this.bufferManager.releaseBuffer(d.buffer, d.size, d.usage));
```
Comment (axinging) [inline tfjs-backend-webgpu/src/backend_webgpu.ts:197]: LGTM.
Maybe we can merge releaseResource and releaseBuffer in the next change.

Comment (gyagp) [reply tfjs-backend-webgpu/src/backend_webgpu.ts:193]: releaseResource includes the buffer and texture management, so they don't need to be merged.

File: tfjs-backend-webgpu/src/backend_webgpu.ts (hunk 4 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/backend_webgpu.ts]: done

File: tfjs-backend-webgpu/src/kernels/FromPixels.ts (hunk 3 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/kernels/FromPixels.ts]: done

File: tfjs-backend-webgpu/src/backend_webgpu.ts (hunk 2 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/backend_webgpu.ts]: done

**Comment by gyagp** [comment]
@xhcao Will you take another look at this PR?

**Review by xhcao — APPROVED** [review]
LGTM, thanks.

**Comment by hujiajie** [comment]
Has anyone encountered this error after the change?

```
Chrome 105.0.0.0 (Windows 10) webgpu backend webgpu webgpu {"WEBGPU_CPU_FORWARD":false} should behave well if WEBGPU_USER_IMPORT is true or false FAILED
 Error: The video element has not loaded data yet. Please wait for `loadeddata` event on the <video> element.
     at fromPixels_ (tfjs-core/src/ops/browser.ts:115:13 <- tfjs/tfjs-backend-webgpu/src/tfjs-backend-webgpu_test_bundle.js:5941:15)
     at Object.<anonymous> (tfjs-core/src/ops/browser.ts:279:10 <- tfjs/tfjs-backend-webgpu/src/tfjs-backend-webgpu_test_bundle.js:6021:14)
     at <Jasmine>
     at tfjs/tfjs-backend-webgpu/src/tfjs-backend-webgpu_test_bundle.js:52:64
     at <Jasmine>
     at __async (tfjs-backend-webgpu/src/tfjs-backend-webgpu_test_bundle.js:36:12)
     at Object.fromPixelsAsync (tfjs-backend-webgpu/src/tfjs-backend-webgpu_test_bundle.js:6004:12)
     at ../../tfjs-backend-webgpu/src/backend_webgpu_test.ts:248:36 <- tfjs/tfjs-backend-webgpu/src/tfjs-backend-webgpu_test_bundle.js:84990:47
     at <Jasmine>
     at fulfilled (tfjs-backend-webgpu/src/tfjs-backend-webgpu_test_bundle.js:39:27)
```

**Comment by hujiajie** [comment]
I did more tests today:

1. the issue seems to be reproducible every time on UHD 620
2. the issue happens occasionally (maybe <10%) on UHD 630
3. the issue was not reproduced if I disabled hardware video decoding on UHD 620

**Comment by gyagp** [comment]
I suggest to open a separate issue for this. Like we discussed in the meeting, can you report the readyState? The Error triggered in TFJS is not necessary and wrong, maybe we can simply remove it.

**Pull request merged**